src/mono/mono/mini/aot-compiler.c

```diff
@@ -14473,11 +14710,86 @@ mono_compile_assembly (MonoAssembly *ass, guint32 opts, const char *aot_options,
 	}
 #endif
 
-	if (acfg->aot_opts.direct_pinvoke && !acfg->aot_opts.static_link) {
-		aot_printerrf (acfg, "The 'direct-pinvoke' AOT option also requires the 'static' AOT option.\n");
+	if ((is_direct_pinvoke_enabled (acfg)) && !acfg->aot_opts.static_link) {
+		aot_printerrf (acfg, "The 'direct-pinvoke' flag, 'direct-pinvokes', and 'direct-pinvoke-lists' AOT options also require the 'static' AOT option.\n");
+		return 1;
+	}
+	if (acfg->aot_opts.direct_pinvoke && (acfg->aot_opts.direct_pinvokes || acfg->aot_opts.direct_pinvoke_lists)) {
+		aot_printerrf (acfg, "The 'direct-pinvoke' flag trumps specified 'direct-pinvokes' and 'direct-pinvoke-lists' arguments. Unset either the flag or the specific direct pinvoke arguments.\n");
 		return 1;
 	}
 
+	gboolean added_direct_pinvoke = TRUE;
+	if (acfg->aot_opts.direct_pinvokes) {
+		for (GList *l = acfg->aot_opts.direct_pinvokes; l; l = l->next) {
+			added_direct_pinvoke = process_specified_direct_pinvokes (acfg, (char*)l->data);
+			if (!added_direct_pinvoke)
+				return 1;
+		}
+	}
+	if (acfg->aot_opts.direct_pinvoke_lists) {
+		for (GList *l = acfg->aot_opts.direct_pinvoke_lists; l; l = l->next) {
+			char *direct_pinvoke_list = (char*)l->data;
+			FILE *direct_pinvoke_list_file = fopen (direct_pinvoke_list, "r");
```

Comment:
If you plan to read complete file into memory, then all this could be replaced with call to g_file_get_contents.

Reply:
That function doesn't explicitly handle OOM checks and since it uses g_malloc that will hit a g_error, `g_error ("Could not allocate %i bytes", x)`,  if it encounters an OOM situation it will abort the process and never return back. I think the caller of g_file_get_contents should assume that it has content on success and if the implementation of it changes or if g_malloc change behavior, it should check for OOM scenario and return back FALSE when failing.